subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/serialization/HierarchicalNameSerializer.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.tasks.compile.incremental.serialization;
+
+import com.google.common.base.CharMatcher;
+import com.google.common.collect.BiMap;
+import com.google.common.collect.HashBiMap;
+import com.google.common.collect.Interner;
+import org.gradle.internal.serialize.AbstractSerializer;
+import org.gradle.internal.serialize.Decoder;
+import org.gradle.internal.serialize.Encoder;
+
+import java.io.IOException;
+
+/**
+ * Efficiently serializes hierarchical names, like Java class names or relative paths of resources.
+ * Splits names into prefixes and suffixes along package separators, inner class separators, file separators and camel case borders.
+ * Reuses these prefixes and suffixes to efficiently store names or parts of names it has seen before.
+ *
+ * This class is stateful. Use a new one for each serialization/deserialization attempt.
+ */
+public class HierarchicalNameSerializer extends AbstractSerializer<String> {
+
+    private final Interner<String> interner;
+    private final BiMap<Integer, String> namesById = HashBiMap.create();
+    private final CharMatcher separatorMatcher;
+
+    public HierarchicalNameSerializer(Interner<String> interner) {
+        this.interner = interner;
+        this.separatorMatcher = CharMatcher.anyOf(".$/").or(CharMatcher.inRange('A', 'Z'));
```

Comment:
This is pretty much the only new thing here - Splitting on camel case and slash in addition to dot and dollar. Saves quite a bit of space for repetitive names like `xxxIntegrationTest`.